sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -142,3 +196,227 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
     result
   }
 }
+
+final class GenericIntArrayData(val primitiveArray: Array[Int]) extends GenericArrayData {
+  override def array(): Array[Any] = primitiveArray.toArray
+
+  override def copy(): GenericIntArrayData = new GenericIntArrayData(primitiveArray.clone())
+
+  override def numElements(): Int = primitiveArray.length
+
+  override def isNullAt(ordinal: Int): Boolean = false
+  override def getInt(ordinal: Int): Int = primitiveArray(ordinal)
+  override def toIntArray(): Array[Int] = {
+    val array = new Array[Int](numElements)
+    System.arraycopy(primitiveArray, 0, array, 0, numElements)
+    array
+  }
+
+  override def equals(o: Any): Boolean = {
+    if (!o.isInstanceOf[GenericIntArrayData]) {
+      return false
+    }
+
+    val other = o.asInstanceOf[GenericIntArrayData]
+    if (other eq null) {
```

Comment:
Nit move this up - seems like the most trivial of these checks. I also feed that a ref check is useful.